### PR TITLE
Increase default test timeout for distributed tests

### PR DESCRIPTION
**Pull request by Flamefire**
When running on clusters the startup time for the subprocesses might be much higher which leads to spurious failures.
So increase this to 500s similar to torch/testing/_internal/distributed/distributed_test.py

FTR: In our tests we even use ridiculously high values such as 3000s to avoid this, so this is even conservative. Maybe an env variable to allow to override this easily would be a good idea?

### CODE REVIEW

**Comment by gcramer23** [comment]
>When running on clusters the startup time for the subprocesses might be much higher which leads to spurious failures.

@Flamefire Have there been failures due to startup time? Can you provide a link?

also I checked https://github.com/pytorch/pytorch/blob/f4967d68f559b085d5a7cc48e1a0eacfaba82ae4/torch/testing/_internal/distributed/distributed_test.py#L120 the default timeout is set to 300.

There is a customized timeout set to 500 "test_DistributedDataParallel". It isn't used in the test file.
https://github.com/pytorch/pytorch/blob/f4967d68f559b085d5a7cc48e1a0eacfaba82ae4/torch/testing/_internal/distributed/distributed_test.py#L121

The change will also make the default timeout larger than the TIMEOUT_OVERRIDE
https://github.com/pytorch/pytorch/blob/b1d17bc55f250383e5814a6ee6c7d11e52ef9c62/torch/testing/_internal/common_distributed.py#L206 so that would need to be fixed as well. 

I would increase the timeout to 300 before increasing to 500 unless it needs to be 500.

**Review by gcramer23 — COMMENTED** [review]
>When running on clusters the startup time for the subprocesses might be much higher which leads to spurious failures.

@Flamefire Have there been failures due to startup time? Can you provide a link?

also I checked https://github.com/pytorch/pytorch/blob/f4967d68f559b085d5a7cc48e1a0eacfaba82ae4/torch/testing/_internal/distributed/distributed_test.py#L120 the default timeout is set to 300.

There is a customized timeout set to 500 "test_DistributedDataParallel". It isn't used in the test file.
https://github.com/pytorch/pytorch/blob/f4967d68f559b085d5a7cc48e1a0eacfaba82ae4/torch/testing/_internal/distributed/distributed_test.py#L121

The change will also make the default timeout larger than the TIMEOUT_OVERRIDE
https://github.com/pytorch/pytorch/blob/b1d17bc55f250383e5814a6ee6c7d11e52ef9c62/torch/testing/_internal/common_distributed.py#L206 so that would need to be fixed as well. 

I would increase the timeout to 300 before increasing to 500 unless it needs to be 500.

**Comment by Flamefire** [comment]
I don't have a link, this is from running the tests on our cluster. One of such failures looks like this:

```
ERROR: test_grad_layout_1devicemodule_1replicaperprocess (__main__.DistributedDataParallelTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/easybuild-tmp/eb-6LtOCS/tmp0oQ5Pe/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py", line 267, in wrapper
    self._join_processes(fn)
  File "/tmp/easybuild-tmp/eb-6LtOCS/tmp0oQ5Pe/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py", line 384, in _join_processes
    self._check_return_codes(elapsed_time)
  File "/tmp/easybuild-tmp/eb-6LtOCS/tmp0oQ5Pe/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py", line 425, in _check_return_codes
    raise RuntimeError('Process {} terminated or timed out after {} seconds'.format(i, elapsed_time))
RuntimeError: Process 0 terminated or timed out after 100.0240318775177 seconds
```

As for your comments: 

- There seem to be multiple default timeouts. The one you quoted (DEFAULT_TIMEOUT and CUSTOMIZED_TIMEOUT) appear to be completely unused, so I guess they fell victim to some refactoring
- I know its larger than the override, but it still makes sense: That test doesn't need such a big timeout but seemingly some others do. I tested 300 successfully on our system but I think it is useful to have some leeway. Your call
- I also noticed that e.g. the CUDA 10.1 build succeeds with the default values while the CUDA 11.1 build times out on 100s

**Comment by gcramer23** [comment]
@Flamefire, I believe the default timeout should be the minimum timeout, and overrides should only be increasing the timeout. If 300 is working, then I see the reason to increase the default timeout to 500. 300 could already be providing some leeway if the tests only require 105.

**Comment by Flamefire** [comment]
Alright, done

**Review by gcramer23 — APPROVED** [review]

**Comment by gcramer23** [comment]
@jbschlosser I reviewed the pr and approved it. I am not authorized to merge the pr.

File: torch/testing/_internal/common_distributed.py [hunk 1]
```diff
@@ -249,7 +249,7 @@ def create_tcp_store(addr="localhost", world_size=1, is_master=True, timeout=tim
         return c10d.TCPStore(addr, port, world_size, is_master, wait_for_workers=wait_for_workers)
 
 
-TIMEOUT_DEFAULT = 100
+TIMEOUT_DEFAULT = 300
```
Comment (malfet) [inline torch/testing/_internal/common_distributed.py:252]:
```suggestion
TIMEOUT_DEFAULT = int(os.getenv('DISTRIBUTED_TESTS_DEFAULT_TIMEOUT', '100'))
```

**Review by malfet — CHANGES_REQUESTED** [review]
Default timeout should be kept as low as possible (i.e. at current value, unless there is an an evidence that it causes false positive failures in our CI), but there are no harm in making it overridable by environment variable.

File: torch/testing/_internal/common_distributed.py (hunk 1 above)
Comment (Flamefire) [reply torch/testing/_internal/common_distributed.py:252]: I'm not opposed to that, as it is an improvement over having to maintain a patch.
However: Shouldn't unit tests always succeed and not cause false positive failures, even when they are run in a different (but non-arcane) environment/CI?

**Comment by Flamefire** [comment]
Not stale but needs rebasing...

**Pull request closed without merging**